Clockwork Patch III

Xhuis opened the pull request:
:cl: Xhuis
tweak: Clockwork marauders can now smash walls and tables.
tweak: Servants are no longer turned into harvesters by Nar-Sie but instead take heavy brute damage and begin to bleed. Similarly, cultists are no longer converted by Ratvar but instead take heavy fire damage and are set ablaze.
rscadd: Added clockwork reclaimers. Ghosts can now click on Ratvar to become clockwork reclaimers - small constructs capable of forcefully converting those that still resist Ratvar's reign. Alt-clicking a valid target will allow the reclaimer to leap onto the head of any non-servant, latching onto their head and converting the target if they are able. The target will be outfitted with an unremovable, acid-proof hat version of the reclaimer. The reclaimer will be able to speak as normal and attack its host and anything nearby. If the reclaimer's host goes unconscious or dies, it will leap free.
rscadd: Added pinion airlocks. These airlocks can only be opened by servants and conventional deconstruction will not work on them. They can be created by proselytizing a normal airlock or from Ratvar.
rscadd: Added ratvarian windows. They are created when Ratvar twists the form of a normal window, and there are both single-direction and full-tile variants.
bugfix: Non-servants can no longer use clockwork proselytizers or mending motors.
bugfix: Servants are now deconverted properly.
/ :cl:

Dropped the post-merge since it isn't really just afterwards anymore.
- Clockwork marauders can now smash racks and tables.
- Added clockwork reclaimers. These tiny robots are the Ratvarian equivalent of harvesters - ghosts can reach out to Ratvar to become one. Although fragile and weak, these robots can alt-click on a valid human to leap onto their head and turn them into servants. The reclaimers will leave their host if they die.
- Ratvar now converts airlocks into pinion airlocks, which can only be opened by servants and have a unique deconstruction (screwdriver, wrench, crowbar).
- Servants are no longer turned into harvesters by Nar-Sie, instead taking a huge amount of brute damage and bleeding profusely. The adverse is true for Ratvar, dealing a huge amount of fire damage and igniting the mob.
- Fixes #18023 
- Fixes #18162 
- Fixes #18163

Commit:
Clockwork post-merge patch 3

Commit:
I'M SORRY

RemieRichards commented on code/game/gamemodes/clock_cult/clock_mobs.dm:
```diff
@@ -276,11 +276,99 @@
+		pixel_y += i
+		sleep(2)
+	if(prob(1))
+		icon_state = "GOTTA_STEP_IT_UP" //thanks, man_shroom
```
Absolutely not.

Commit:
Removes fun

ChangelingRain commented:
>reclaimer charge sprite has only one direction

aaaaaaaa

ChangelingRain commented:
No it isn't!

Xhuis commented:
fiiine

RemieRichards commented on code/game/gamemodes/clock_cult/clock_mobs.dm:
```diff
@@ -276,11 +276,99 @@
+		pixel_y += i
+		sleep(2)
+	if(prob(1))
+		maptext = "GOTTA STEP IT UP" //thanks man_shroom
```
No, not this either.

Commit:
ugh

Anonmare commented:
So can Servants make them by themselves? Would seem like something a Clockwork Proselytizer could convert.

Xhuis commented:
I'll probably change it before the merge, but servants can't currently make them.

Commit:
Proselytizer tweaks

Xhuis commented:
__New in this commit:__
* Clockwork wall cost changed to 5 liquified replicant alloy (down from 10).
* Clockwork floor cost changed to 1 liquified replicant alloy (down from 5).
* Clockwork proselytizers can now make normal airlocks into pinion airlocks for 5 liquified replicant alloy.

Commit:
Touchups and polish

Xhuis commented:
__New in this commit:__
* Split up `clock_objects.dm` into `clock_items.dm` and `clock_structures.dm` as requested by @phil235 
* Added directional sprites for the charging reclaimer (not posted here; they're just motions blurs)
* Targets dominated by reclaimers now lose their reclaimer hat once the reclaimer leaves, although they remain servants

Commit:
Changes some small stuff

Commit:
Now it compiles!

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_unsorted.dm:
```diff
@@ -97,6 +97,97 @@
+#define GEAR_UNFASTENED 2
+#define GEAR_LOOSE 3
+
+/obj/machinery/door/airlock/clockwork
```
Add a New() for these with this so you get a cool spawn animation(this may require updating your code so it can compile)
    
var/turf/T = get_turf(src)
PoolOrNew(/obj/effect/overlay/temp/ratvar/door, T)
PoolOrNew(/obj/effect/overlay/temp/ratvar/beam/door, T)

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_unsorted.dm:
```diff
@@ -97,6 +97,97 @@
+
+/obj/machinery/door/airlock/clockwork/attackby(obj/item/I, mob/living/user, params)
+	if(!attempt_construction(I, user))
+		..()
```
return ..()

ChangelingRain commented:
Here's door sprites updated to work with our door system http://puu.sh/pchkM/abbe25069a.zip

optimumtact commented:
tagged dnm until you say so

Xhuis commented:
Thanks.

phil235 commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,755 @@
+		health = max(0, health - I.force)
+		if(!health)
+			destroyed()
+			return 0
```
I refactored attackby so you could all attackable structures and machines would use the same code to handle this... Look at item_attack.dm . See how grille or windows handle it now.

phil235 commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,755 @@
+		qdel(S)
+		qdel(src)
+		return 1
+	..()
```
else return ..()

phil235 commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,755 @@
+			continue
+		C.health = min(C.health + 10, C.max_health)
+		if(uses_alloy)
+			stored_alloy = max(0, stored_alloy - 1)
```
Instead of checking everything in range(5, src) FOUR TIMES, you could search for atom/movable in that range just once and then check if(istype(anima) else if(marauder) else if(silicon)...

phil235 commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,755 @@
+		user.drop_item()
+		qdel(I)
+		return 1
+	..()
```
else return ..()

phil235 commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,755 @@
+		user.drop_item()
+		qdel(I)
+		return 1
+	..()
```
else return ..()

phil235 commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,755 @@
+/obj/structure/clockwork/cache/attackby(obj/item/I, mob/living/user, params)
+	if(!is_servant_of_ratvar(user))
+		..()
+		return 0
```
return ..()

phil235 commented on code/modules/mob/living/living.dm:
```diff
@@ -773,6 +773,12 @@ Sorry Giacom. Please don't be mad :(
+		adjustBruteLoss(35)
+		if(reagents)
+			reagents.add_reagent("heparin", 5)
+		return 0
```
is this really supposed to be something happening? clockcult antag meeting narsie?

Anonmare replied:
Yeah

Xhuis replied:
What Anonmare said. I'm probably going to be implementing the cult v. cult mechanics and submode in this PR as well.

Commit:
Merge remote-tracking branch 'remotes/upstream/master' into clockwork_patch_3

Commit:
Optimizations courtesy of @phil235

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,764 @@
+		toggle()
+		return 0
+	for(var/atom/movable/M in range(5, src))
+		switch(M.type)
```
Bad! You need to do istype(thing, type) or you won't get subtypes.
Same for the clockwork proselytizer, now that I check the code.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_ratvar.dm:
```diff
@@ -239,15 +249,30 @@
+	if(!add_servant_of_ratvar(src) && !is_servant_of_ratvar(src))
+		src << "<span class='userdanger'>A blinding light boils you alive! <i>Run!</i></span>"
+		adjustFireLoss(35)
+		adjust_fire_stacks(1)
```
Make sure the mob still exists after dealing damage to it.

Xhuis replied:
See below

ChangelingRain replied:
If it deletes on death, like harvesters do, it'll runtime

ChangelingRain commented on code/modules/mob/living/living.dm:
```diff
@@ -773,6 +773,12 @@ Sorry Giacom. Please don't be mad :(
 /mob/living/narsie_act()
+	if(is_servant_of_ratvar(src) && !stat)
+		src << "<span class='userdanger'>You resist Nar-Sie's influence... but not all of it. <i>Run!</i></span>"
+		adjustBruteLoss(35)
```
Most mobs with reagents won't be deleted after taking damage, but make sure they still exist anyway

Xhuis replied:
The code executes instantly after the check, so it shouldn't be necessary.

ChangelingRain replied:
If it deletes on death it'll runtime

Commit:
Fixes a few issues and tweaks some things

Xhuis commented:
__New in this commit:__
* Fixes #18162 
* Fixes #18163 
* Pinion airlock deconstruction now works correctly
* Proselytizer and mending motor code has been optimized to include subtypes
* Added some more guiding comments to the code
* Added some cultist-only effects on certain scripture

Xhuis commented:
Follow-up commit coming soon

Commit:
Follow-up commit

Xhuis commented:
As promised.
* Ratvarian windows have been added. They cannot be proselytized from normal windows yet but are created by Ratvar. Thanks to @ChangelingRain for sprites, will post shortly.
* It actually compiles now.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_items.dm:
```diff
@@ -571,12 +571,12 @@
 		alloy_cost = 5
 		valid_target = TRUE //Need to change valid_target to 1 or TRUE in each check so that it doesn't return an invalid value
-	else if(istype(src, /turf/open/floor/plating))
+	else if(istype(src, /turf/open/floor))
```
You may want to prevent converting clockwork floors.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_items.dm:
```diff
@@ -571,12 +571,12 @@
 	var/alloy_cost = 0
 	var/valid_target = FALSE //If the proselytizer will actually function on the object
-	if(istype(target, /turf/closed/wall) && !istype(target, /turf/closed/wall/reinforced))
+	if(istype(target, /turf/closed/wall) && !istype(target, /turf/closed/wall/r_wall))
```
Same here, might want to prevent converting the clockwork walls.

ChangelingRain commented:
The fulltile clockwork window is supposed to smooth, you don't seem to have the smoothed sprites.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_unsorted.dm:
```diff
@@ -200,6 +200,22 @@
 
+/obj/structure/window/reinforced/clockwork
+	name = "ratvarian window"
+	desc = "A reinforced yellow-glass window, gently luminous."
```
I actually intended it as a brass pane thin enough to see through. Glass doesn't really fit ratvar, honestly.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_unsorted.dm:
```diff
@@ -200,6 +200,22 @@
+
+/obj/structure/window/reinforced/clockwork/New()
+	..()
+	spawn(1)
```
If you give it some arguments in New() you won't need to use a spawn(1) and I'd recommend doing so

Commit:
Some more work or something, idk

Commit:
Sanity check 2

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_items.dm:
```diff
@@ -571,12 +571,12 @@
 	var/alloy_cost = 0
 	var/valid_target = FALSE //If the proselytizer will actually function on the object
-	if(istype(target, /turf/closed/wall) && !istype(target, /turf/closed/wall/r_wall))
+	if(istype(target, /turf/closed/wall) && !target.type in list(/turf/closed/wall/r_wall, /turf/closed/wall/clockwork))
```
USE ISTYPE YOU BUTT this won't catch rusted rwalls

Xhuis commented:
__New in this commit:__
* Stopped being stubborn and moved all of the clockwork variants of stuff in `clock_unsorted.dm` to the files they should be in anyway.
* Added Ratvarian windows (sprites by @ChangelingRain) that Ratvar will create from normal windows.
* Ratvar and the celestial gateway are no longer dense.
* Introduced additional sanity checking in `narsie_act` and `ratvar_act` for mobs

ChangelingRain commented on code/game/objects/structures/window.dm:
```diff
@@ -490,3 +490,27 @@
+	desc = "A paper-thin pane of translucent yet reinforced brass."
+	icon = 'icons/obj/clockwork_objects.dmi'
+	icon_state = "clockwork_window_single"
+	debris = list(/obj/item/clockwork/component/vanguard_cogwheel)
```
This will not work; if you want custom debris, you need to delete the debris spawned in New()'s ..() and add this to the debris.

ChangelingRain commented on code/game/turfs/simulated/floor/misc_floor.dm:
```diff
@@ -110,4 +110,40 @@
+		user.visible_message("<span class='notice'>[user] pries up [src]!</span>", "<span class='notice'>You pry up [src], destroying it!</span>")
+		make_plating()
+		return 1
+	..()
```
return ..()

ChangelingRain commented on code/game/turfs/simulated/walls_misc.dm:
```diff
@@ -32,6 +32,55 @@
+		user.visible_message("<span class='notice'>[user] breaks apart [src]!</span>", "<span class='notice'>You break apart [src]!</span>")
+		break_wall()
+		return 1
+	..()
```
return ..()

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_items.dm:
```diff
@@ -556,17 +571,21 @@
-			new_obj_type = /turf/open/floor/clockwork
-			alloy_cost = 5
-			valid_target = TRUE
+	if(istype(target, /turf/closed/wall) && !target.type in list(/turf/closed/wall/r_wall, /turf/closed/wall/clockwork))
```
use istype so you catch subtypes!!!

Xhuis replied:
__ME SMART__

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,768 @@
+		user << "<span class='warning'>[src] is empty!</span>"
+		return 0
+	var/component_to_withdraw = input(user, "Choose a component to withdraw.", name) as null|anything in possible_components
+	if(!user || !user.canUseTopic(src) || !component_to_withdraw)
```
You need to also check if there's any of the chosen component left, though right now resource generation is bonkers so it doesn't matter too much.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,768 @@
+		damage_per_tick = 10
+		sight_range = 5
+	if(target)
+		if(target.stat || get_dist(get_turf(src), get_turf(target)) > sight_range)
```
Should also check if the target is a servant, ie; it got converted while it was attacking

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,768 @@
+		else
+			target.adjustFireLoss(!iscultist(target) ? damage_per_tick : damage_per_tick * 2) //Nar-Sian cultists take additional damage
+			if(ratvar_awakens)
+				target.adjust_fire_stacks(damage_per_tick)
```
Make sure the target still exists by this point, if it's a construct it may have died and been deleted already.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,768 @@
+	clockwork_desc = "A sigil that will soon erupt and smite any unenlightened nearby."
+	icon = 'icons/effects/96x96.dmi'
+	pixel_x = -30
+	pixel_y = -30
```
Both of these should be -32

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,768 @@
+		var/mob/M = A
+		M.forceMove(get_turf(linked_gateway))
+	else
+		A.loc = get_turf(linked_gateway)
```
It doesn't matter what type it is, you should use forceMove() for both

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,768 @@
+				stored_alloy = max(0, stored_alloy - 2)
+		else if(istype(M, /mob/living/simple_animal/hostile/clockwork_marauder))
+			var/mob/living/simple_animal/hostile/clockwork_marauder/E = M
+			if(E.health == E.maxHealth || E.stat)
```
Should this not also check fatigue? IE; if((E.health == E.maxHealth && !fatigue) || E.stat)

Commit:
More stuff, no conflict fix yet

Commit:
Uhh?

Xhuis commented:
Almost all code is optimized now. Conflicts haven't been fixed yet; there's something I need to do.

ChangelingRain commented:
All you need to do is resolve using yours and add the changes from #18166 to the walls

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,764 @@
+				stored_alloy = max(0, stored_alloy - 2)
+		else if(istype(M, /mob/living/simple_animal/hostile/clockwork_marauder))
+			var/mob/living/simple_animal/hostile/clockwork_marauder/E = M
+			if(E.health == E.maxHealth || E.stat || !E.fatigue)
```
you want if((E.health == E.maxHealth && !E.fatigue) || E.stat), or it'll stop when it's EITHER at max health OR has no fatigue, which under normal conditions will be true.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,764 @@
+		return ..()
+
+/obj/structure/clockwork/mending_motor/proc/toggle(mob/living/user)
+	active = !active
```
Shouldn't you make sure the user(if there is one) is a servant before toggling it, so nonservants can't mess with it?

Commit:
Two more small fixes

Commit:
Conflicts I

Commit:
Conflicts II

Commit:
Conflicts III

Xhuis commented:
Conflicts fixed and code tweaked. This should be merge-ready now.

Xhuis commented:
Odd, just worked on my end. Hold on...

Commit:
YOU CAN'T HIDE FROM ME

Xhuis commented:
After some thermal scanning we found a rogue whitespace sneaking around in the facility. He's been eliminated and this is now merge-ready for real this time.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,765 @@
+		return ..()
+
+/obj/structure/clockwork/mending_motor/proc/toggle(mob/living/user)
+	if(!user || !is_servant_of_ratvar(user))
```
er, don't you need && here to replicate previous 'toggle off when out of alloy/when spawned' behavior?

Xhuis replied:
I use that style because if `user` doesn't exist in the first place it ends immediately but if there is then it goes on to check if they're not a servant.

Commit:
I can code guys I swear

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -462,9 +462,10 @@
+	if(user)
+		if(!is_servant_of_ratvar(user))
+			return 0
+		user.visible_message("<span class='notice'>[user] [active ? "en" : "dis"]ables [src].</span>", "<span class='brass'>You [active ? "en" : "dis"]able [src].</span>")
```
you need to swap the `[active ? "en" : "dis"]`es to `[active ? "dis" : "en"]` because this message is before the active = !active toggle

phil235 commented on code/game/gamemodes/clock_cult/clock_structures.dm:
```diff
@@ -0,0 +1,766 @@
+	qdel(src)
+	return 1
+
+/obj/structure/clockwork/proc/damaged(mob/living/user, obj/item/I, amount, type)
```
"type" is also a var name, it's risky to reuse a var name for a proc argument.

Commit:
Done and done

Xhuis commented:
That should be all, then.
